core/src/main/java/zipkin/reporter/BatchReporter.java

```diff
@@ -0,0 +1,209 @@
+ */
+public final class BatchReporter implements Reporter, Flushable, Component {
+
+  public static BatchReporter create(Sender sender) {
+    return builder().sender(sender).build();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    SpanEncoder spanEncoder = SpanEncoder.THRIFT;
+    Sender sender = null;
+    ReporterMetrics metrics = ReporterMetrics.NOOP_METRICS;
+    int flushInterval = 1;
+    private int maxQueueDepth = 1000;
+
+    Builder() {
+    }
+
+    /**
+     * As spans are reported, they are {@link SpanEncoder#encode(Span)} serialized using this
+     * component. Defaults to {@link SpanEncoder#THRIFT thrift} which is the least overhead.
+     */
+    public Builder spanEncoder(SpanEncoder spanEncoder) {
+      this.spanEncoder = checkNotNull(spanEncoder, "spanEncoder");
+      return this;
+    }
+
+    /**
+     * Once spans are drained, they are {@link Sender#sendSpans(List, Callback)} using this
+     * component.
+     */
+    public Builder sender(Sender sender) {
+      this.sender = checkNotNull(sender, "sender");
+      return this;
+    }
+
+    /**
+     * Aggregates and reports reporter metrics to a monitoring system. Should be {@link
+     * ReporterMetrics#forTransport(String) scoped to this transport}.  Defaults to no-op.
+     */
+    public Builder metrics(ReporterMetrics metrics) {
+      this.metrics = checkNotNull(metrics, "metrics");
+      return this;
+    }
+
+    /** Default 1 second. 0 implies spans are {@link #flush() flushed} externally. */
+    public Builder flushInterval(int flushInterval) {
+      this.flushInterval = flushInterval;
+      return this;
+    }
+
+    /** Maximum backlog of spans reported vs sent. Default 1000 */
+    public Builder maxQueueDepth(int maxQueueDepth) {
+      this.maxQueueDepth = maxQueueDepth;
+      return this;
+    }
+
+    public BatchReporter build() {
+      return new BatchReporter(this);
+    }
+  }
+
+  final Logger logger;
+  final Sender sender;
+  final SpanEncoder spanEncoder;
+  final ReporterMetrics metrics;
+  final BlockingDeque<byte[]> pending;
+  @Nullable // for manually flushing
+  final Flusher flusher;
+
+  BatchReporter(Builder builder) {
+    this(builder, Logger.getLogger(BatchReporter.class.getName()));
+  }
+
+  BatchReporter(Builder builder, Logger logger) { // Visible for testing
+    this.sender = checkNotNull(builder.sender, "sender");
+    this.logger = logger;
+    this.spanEncoder = checkNotNull(builder.spanEncoder, "spanEncoder");
+    this.metrics = builder.metrics == null ? ReporterMetrics.NOOP_METRICS : builder.metrics;
+    this.pending = new LinkedBlockingDeque<>(builder.maxQueueDepth);
+    this.flusher = builder.flushInterval > 0 ? new Flusher(this, builder.flushInterval,
+        "Flusher(" + sender + ")") : null;
+  }
+
+  @Override
+  public void report(Span span) {
+    metrics.incrementSpans(1);
+    if (!pending.offer(spanEncoder.encode(span))) {
```

Comment:
Good thoughts. So I'd expect the queue (or buffer) to always be bounded.
What do do when it is full (besides dropping) is a good question. There's
another matter which is whether to drop a span too large to send. For
example kafka has a max message size which people have (usually
erroneously) exceeded.

You are right that we can rejig things to allow a flush ahead of schedule
(based on size) though that would more of a precaution than a cure of a
full queue.

Lets revisit this again on next round.. thanks for the notes!

On 22 Aug 2016 10:17 am, "Nic Munroe" <notifications@github.com> wrote:

> In core/src/main/java/zipkin/reporter/BatchReporter.java
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7#discussion_r75610994>
> :
>
> > +  }
> > +
> > +  BatchReporter(Builder builder, Logger logger) { // Visible for testing
> > +    this.sender = checkNotNull(builder.sender, "sender");
> > +    this.logger = logger;
> > +    this.spanEncoder = checkNotNull(builder.spanEncoder, "spanEncoder");
> > +    this.metrics = builder.metrics == null ? ReporterMetrics.NOOP_METRICS : builder.metrics;
> > +    this.pending = new LinkedBlockingDeque<>(builder.maxQueueDepth);
> > +    this.flusher = builder.flushInterval > 0 ? new Flusher(this, builder.flushInterval,
> > +        "Flusher(" + sender + ")") : null;
> > +  }
> > +
> > +  @Override
> > +  public void report(Span span) {
> > +    metrics.incrementSpans(1);
> > +    if (!pending.offer(spanEncoder.encode(span))) {
>
> Any thought to alternate behavior besides simply dropping the span? You
> could have it immediately flush the sender with whatever you have even if
> it's not the regularly scheduled sending time. Maybe the ability to have it
> auto-flush like that when it's 80% full or something?
>
> These alternate behaviors could add some non-trivial complexity, so maybe
> it's not worth it at this point (or ever depending on if people would
> actually use it). Something to think about though.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7/files/cdafba588b33e42a6a8a148bb733dd0b9bae5806#r75610994>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD613UKlayASO7K7RnKI9yeJHVddg2Dks5qiQbFgaJpZM4JpQw9>
> .
>